test/functional/feature_onlynet.py

```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021-2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test -onlynet
+"""
+
+from test_framework.basic_server import (
+    BasicServer,
+    tor_control
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+)
+
+
+class OnlynetTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.extra_args = [
+            ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1'],
+            ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-onion=127.0.0.1:9050'],
+            ['-onlynet=ipv4', '-torcontrol=127.0.0.1:19051', '-listenonion=1', '-proxy=127.0.0.1:9050'],
+        ]
+        self.num_nodes = len(self.extra_args)
+        self.tor_control_server = BasicServer(bind=('127.0.0.1', 19051),
+                                              response_generator=tor_control)
+
+
+    def run_test(self):
+        def networks_dict(d):
+            r = {}
+            for x in d['networks']:
+                r[x['name']] = x
+            return r
+
+        # Would fail without https://github.com/bitcoin/bitcoin/pull/22651
+        n0 = networks_dict(self.nodes[0].getnetworkinfo())
+        assert_equal(n0['onion']['reachable'], False)
+
+        # Passes with or without https://github.com/bitcoin/bitcoin/pull/22651,
+        # just to ensure behavior does not change unintended.
+        n1 = networks_dict(self.nodes[1].getnetworkinfo())
+        assert_equal(n1['onion']['reachable'], True)
+
+        # Passes with or without https://github.com/bitcoin/bitcoin/pull/22651,
+        # just to ensure behavior does not change unintended.
+        n2 = networks_dict(self.nodes[2].getnetworkinfo())
+        assert_equal(n2['onion']['reachable'], True)
```

Comment:
Here is a proposed simplification using a dict comprehension in `networks_dict` and a loop for the assertions:
```diff
+    def networks_dict(self, node):
+        info = self.nodes[node].getnetworkinfo()
+        return {network["name"]: network for network in info['networks']}
 
     def run_test(self):
-        def networks_dict(d):
-            r = {}
-            for x in d['networks']:
-                r[x['name']] = x
-            return r
-
-        # Would fail without https://github.com/bitcoin/bitcoin/pull/22651
-        n0 = networks_dict(self.nodes[0].getnetworkinfo())
-        assert_equal(n0['onion']['reachable'], False)
-
-        # Passes with or without https://github.com/bitcoin/bitcoin/pull/22651,
-        # just to ensure behavior does not change unintended.
-        n1 = networks_dict(self.nodes[1].getnetworkinfo())
-        assert_equal(n1['onion']['reachable'], True)
-
-        # Passes with or without https://github.com/bitcoin/bitcoin/pull/22651,
-        # just to ensure behavior does not change unintended.
-        n2 = networks_dict(self.nodes[2].getnetworkinfo())
-        assert_equal(n2['onion']['reachable'], True)
+        for node, reachable in [[0, False], [1, True], [2, True]]:
+            assert_equal(self.networks_dict(node)['onion']['reachable'], reachable)
```
Could perhaps add a comment describing why we expect the first node to not be reachable.

Reply:
Done something similar. I find `return {network["name"]: network for network in info["networks"]}` extremely hard to read.

Reply:
Comprehensions seem idiomatic/pythonic and the most-used way to do it in this codebase too (though I agree that a shorter `n` or `net` for the one line would probably be better than `network`), but I like your new version too. Nice and simple.